src/3d/qgstiledscenechunkloader_p.cpp

```diff
@@ -0,0 +1,325 @@
+{
+  QgsDataSourceUri dsUri;
+  dsUri.setEncodedUri( baseUri );
+
+  const QString tileSetUri = dsUri.param( QStringLiteral( "url" ) );
+  if ( !tileSetUri.isEmpty() )
+  {
+    QUrl base( tileSetUri );
+    uri = base.resolved( uri ).toString();
+  }
+  else
+  {
+    // local files
+    if ( uri.startsWith( "./" ) )
+    {
+      uri.replace( "./", baseUri );
+    }
+    else if ( QFileInfo( uri ).isRelative() )
+    {
+      uri = baseUri + uri;
+    }
+  }
+  return uri;
+}
+
+static QgsChunkNodeId nodeIdFromUuid( const QString &uuid )
+{
+  QgsChunkNodeId nodeId;
+  QUuid nodeUuid = QUuid::fromString( uuid );
+  Q_ASSERT( !nodeUuid.isNull() );
+  Q_ASSERT( sizeof( QUuid ) == 16 && sizeof( QgsChunkNodeId ) == 16 );
+  memcpy( reinterpret_cast<char *>( &nodeId ), reinterpret_cast<const char *>( &nodeUuid ), 16 );
+  return nodeId;
+}
+
+static QString uuidFromNodeId( const QgsChunkNodeId &nodeId )
+{
+  QUuid nodeUuid;
+  Q_ASSERT( sizeof( QUuid ) == 16 && sizeof( QgsChunkNodeId ) == 16 );
+  memcpy( reinterpret_cast<char *>( &nodeUuid ), reinterpret_cast<const char *>( &nodeId ), 16 );
+  return nodeUuid.toString();
+}
+
+///
+
+QgsTiledSceneChunkLoader::QgsTiledSceneChunkLoader( QgsChunkNode *node, const QgsTiledSceneChunkLoaderFactory *factory, const QgsTiledSceneTile &t )
+  : QgsChunkLoader( node ), mFactory( factory ), mTile( t )
+{
+  qDebug() << "chunk loader " << node->tileId().text();
+
+  // start async (actually just do deferred finish()!)
+  // TODO: loading in background thread
+  QMetaObject::invokeMethod( this, "finished", Qt::QueuedConnection );
+}
+
+Qt3DCore::QEntity *QgsTiledSceneChunkLoader::createEntity( Qt3DCore::QEntity *parent )
+{
+  qDebug() << "create entity!" << mNode->tileId().text();
+
+  // we do not load tiles that are too big - at least for the time being
+  // the problem is that their 3D bounding boxes with ECEF coordinates are huge
+  // and we are unable to turn them into planar bounding boxes
+  if ( hasLargeBounds( mTile ) )
+  {
+    return new Qt3DCore::QEntity( parent );
+  }
+
+  QString uri = mTile.resources().value( QStringLiteral( "content" ) ).toString();
+  if ( uri.isEmpty() )
+  {
+    // nothing to show for this tile
+    // TODO: can we skip loading it at all?
+    return new Qt3DCore::QEntity( parent );
+  }
+
+  uri = resolveUri( uri, mFactory->mRelativePathBase );
+
+  qDebug() << "loading: " << uri;
+  QByteArray content = mFactory->mIndex.retrieveContent( uri );
+
+  if ( content.isEmpty() )
+  {
+    // the request probably failed
+    // TODO: how can we report it?
+    qDebug() << "retrieveContent returned no content";
+    return new Qt3DCore::QEntity( parent );
+  }
+
+  QByteArray gltfData;
+  if ( content.startsWith( QByteArray( "b3dm" ) ) )
+  {
+    QBuffer buffer( &content );
+    buffer.open( QIODevice::ReadOnly );
+    gltfData = QgsCesiumUtils::extractGltfFromB3dm( buffer );
+  }
+  else if ( content.startsWith( QByteArray( "glTF" ) ) )
+  {
+    gltfData = content;
+  }
```

Comment:
We could move this to QgsCesiumUtils as `extractGltfFromData` (would be useful in 2d as well)